source/common/tracing/opentracing_driver_impl.cc

```diff
@@ -0,0 +1,168 @@
+    }
+  }
+};
+} // namespace
+
+OpenTracingSpan::OpenTracingSpan(bool use_single_header_propagation, bool use_tracer_propagation,
+                                 std::unique_ptr<opentracing::Span>&& span)
+    : use_single_header_propagation_(use_single_header_propagation),
+      use_tracer_propagation_(use_tracer_propagation), span_(std::move(span)) {}
+
+void OpenTracingSpan::finishSpan() { span_->Finish(); }
+
+void OpenTracingSpan::setOperation(const std::string& operation) {
+  span_->SetOperationName(operation);
+}
+
+void OpenTracingSpan::setTag(const std::string& name, const std::string& value) {
+  span_->SetTag(name, value);
+}
+
+void OpenTracingSpan::injectContext(Http::HeaderMap& request_headers) {
+  if (use_single_header_propagation_) {
+    // Inject the span context using Envoy's single-header format.
+    std::ostringstream oss;
+    const opentracing::expected<void> was_successful =
+        span_->tracer().Inject(span_->context(), oss);
+    if (!was_successful) {
+      ENVOY_LOG(warn, "Failed to inject span context: {}", was_successful.error().message());
+      return;
+    }
+    const std::string current_span_context = oss.str();
+    request_headers.insertOtSpanContext().value(
+        Base64::encode(current_span_context.c_str(), current_span_context.length()));
+  }
+
+  if (use_tracer_propagation_) {
+    // Inject the context using the tracer's standard HTTP header format.
+    const OpenTracingHTTPHeadersWriter writer{request_headers};
+    const opentracing::expected<void> was_successful =
+        span_->tracer().Inject(span_->context(), writer);
+    if (!was_successful) {
+      ENVOY_LOG(warn, "Failed to inject span context: {}", was_successful.error().message());
+      return;
+    }
+  }
+}
+
+SpanPtr OpenTracingSpan::spawnChild(const Config&, const std::string& name, SystemTime start_time) {
+  std::unique_ptr<opentracing::Span> ot_span = span_->tracer().StartSpan(
+      name, {opentracing::ChildOf(&span_->context()), opentracing::StartTimestamp(start_time)});
+  if (ot_span == nullptr) {
+    return nullptr;
+  }
+  return SpanPtr{new OpenTracingSpan{use_single_header_propagation_, use_tracer_propagation_,
+                                     std::move(ot_span)}};
+}
+
+SpanPtr OpenTracingDriver::startSpan(const Config&, Http::HeaderMap& request_headers,
+                                     const std::string& operation_name, SystemTime start_time) {
+  const bool use_single_header_propagation = this->useSingleHeaderPropagation();
+  const bool use_tracer_propagation = this->useTracerPropagation();
+  const opentracing::Tracer& tracer = this->tracer();
+  std::unique_ptr<opentracing::Span> active_span;
+  std::unique_ptr<opentracing::SpanContext> parent_span_ctx;
+  if (use_single_header_propagation && request_headers.OtSpanContext()) {
+    std::string parent_context = Base64::decode(request_headers.OtSpanContext()->value().c_str());
+    std::istringstream iss(parent_context);
```

Comment:
stringstreams have fairly poor perf characteristics last I checked. Is this required?

Reply:
I'll looking into adding binary propagation directly into/from string-like objects to the OT API.

Reply:
Put in https://github.com/opentracing/opentracing-cpp/pull/42 to add serialization support directly to/from strings.